eureka/src/main/java/com/linecorp/armeria/server/eureka/EurekaUpdatingListener.java

```diff
@@ -0,0 +1,320 @@
+     * Returns a new {@link EurekaUpdatingListenerBuilder} created with the specified {@code eurekaUri}.
+     */
+    public static EurekaUpdatingListenerBuilder builder(String eurekaUri) {
+        return builder(URI.create(requireNonNull(eurekaUri, "eurekaUri")));
+    }
+
+    /**
+     * Returns a new {@link EurekaUpdatingListenerBuilder} created with the specified {@code eurekaUri}.
+     */
+    public static EurekaUpdatingListenerBuilder builder(URI eurekaUri) {
+        return new EurekaUpdatingListenerBuilder(eurekaUri);
+    }
+
+    /**
+     * Returns a new {@link EurekaUpdatingListenerBuilder} created with the specified {@link SessionProtocol}
+     * and {@link EndpointGroup}.
+     */
+    public static EurekaUpdatingListenerBuilder builder(
+            SessionProtocol sessionProtocol, EndpointGroup endpointGroup) {
+        return new EurekaUpdatingListenerBuilder(sessionProtocol, endpointGroup, null);
+    }
+
+    /**
+     * Returns a new {@link EurekaUpdatingListenerBuilder} created with the specified {@link SessionProtocol},
+     * {@link EndpointGroup} and path.
+     */
+    public static EurekaUpdatingListenerBuilder builder(
+            SessionProtocol sessionProtocol, EndpointGroup endpointGroup, String path) {
+        return new EurekaUpdatingListenerBuilder(sessionProtocol, endpointGroup, requireNonNull(path, "path"));
+    }
+
+    private final EurekaWebClient client;
+    private final InstanceInfo instanceInfo;
+    @Nullable
+    private volatile ScheduledFuture<?> heartBeatFuture;
+
+    @Nullable
+    private volatile String appName; // Set when serverStarted is called.
+    private volatile boolean closed;
+
+    /**
+     * Creates a new instance.
+     */
+    EurekaUpdatingListener(EurekaWebClient client, InstanceInfo instanceInfo) {
+        this.client = client;
+        this.instanceInfo = instanceInfo;
+    }
+
+    @Override
+    public void serverStarted(Server server) throws Exception {
+        final InstanceInfo newInfo = fillAndCreateNewInfo(instanceInfo, server);
+
+        try (ClientRequestContextCaptor contextCaptor = Clients.newContextCaptor()) {
+            final HttpResponse response = client.register(newInfo);
+            final ClientRequestContext ctx = contextCaptor.get();
+            response.aggregate().handle((res, cause) -> {
+                if (closed) {
+                    return null;
+                }
+                if (cause != null) {
+                    logger.warn("Failed to register {} to Eureka: {}",
+                                newInfo.getHostName(), client.uri(), cause);
+                    return null;
+                }
+                final ResponseHeaders headers = res.headers();
+                if (headers.status() != HttpStatus.NO_CONTENT) {
+                    logger.warn("Failed to register {} to Eureka: {}. (status: {}, content: {})",
+                                newInfo.getHostName(), client.uri(), headers.status(), res.contentUtf8());
```

Comment:
Don't we need to retry to other eureka servers? Or, does it happen in EurekaWebClient?

Reply:
Add `RetryingClient` if the user didn't specify one. 😉